students/Yue_Ma/Lesson03/mailroom.py

```diff
@@ -60,38 +60,40 @@ def send_a_thank_you_note():
             break
         else:
             print("Not a valid option!!!")
-    return main()
+    return
 
 
 def create_a_report():
     title = {'Donor Name': ['Total Given', 'Num Gifts', 'Average Gift']}
     value_title = list(title.values())
     key_title = list(title.keys())
 
-    # sort
-    donor_list_sort = dict(sorted(donor_list.items(), key=lambda t: t[0]))
-    print(donor_list_sort)
-
     # calculate total donation
-    donation_matrix = []
-    for j in range(0, len(donor_list_sort)):
-        donation_matrix.append([list(donor_list_sort.keys())[j], sum(list(donor_list_sort.values())[j]),
-                                len(list(donor_list_sort.values())[j])])
+    donation_matrix = {}
+    for item in donor_list.keys():
+        donation_matrix[item] = sum(list(donor_list[item]))
+
+    # sort
+    donation_matrix_raw = sorted(donation_matrix.items(), reverse=True, key=lambda t: t[1])
+    donation_matrix_sorted = {}
+    for i in range(len(donation_matrix_raw)):
+        donation_matrix_sorted[donation_matrix_raw[i][0]] = donation_matrix_raw[i][1]
 
+    print(donation_matrix_sorted)
+    # Print the report
     line_title = f'| {key_title[0]:<20}| {value_title[0][0]:<15} | {value_title[0][1]:<10} | {value_title[0][2]:<15}|'
     line_x = '|' + '-' * 69 + '|'
     line_1 = ' ' + '-' * 69 + ' '
 
-    # Print the report
     print(line_1)
     print(line_title)
     print(line_x)
-    for i in range(0, len(donation_matrix)):
-        main_lines = f'| {donation_matrix[i][0]:<20}| {donation_matrix[i][1]:<15.2f} | {donation_matrix[i][2]:<10} ' \
-                     f'| {donation_matrix[i][1] / donation_matrix[i][2]:<15.2f}|'
+    for item in donation_matrix_sorted.keys():
+        main_lines = f'| {item:<20}| {donation_matrix_sorted[item]:<15.2f} | {len(donor_list[item]):<10} ' \
+                     f'| {float(donation_matrix_sorted[item]) / len(donor_list[item]):<15.2f}|'
```

Comment:
here is the simplest approach demonstrated using some sample code:
```
>>> d = {"bob": (100, 200), "jen": (1, 2)}
>>> def _key(data):
...     return sum(data[1])
...
>>> sorted(d.items(), key=_key, reverse=True)
[('bob', (100, 200)), ('jen', (1, 2))]
>>> sorted(d.items(), key=_key, reverse=False)
[('jen', (1, 2)), ('bob', (100, 200))]
```

above code gives you sorted list in a few clear lines of code, then
```
for donor in sorted_donors:
    print("...")
```

in comparison your code right now has two additional loops that are not needed and can make the reader of the code get "lost" when understanding the logic.